components/script/build.rs

```diff
@@ -16,11 +16,17 @@ fn main() {
     let start = Instant::now();
 
     let style_out_dir = PathBuf::from(env::var_os("DEP_SERVO_STYLE_CRATE_OUT_DIR").unwrap());
+    let css_properties_json = style_out_dir.join("css-properties.json");
     let out_dir = PathBuf::from(env::var_os("OUT_DIR").unwrap());
 
+    println!("cargo::rerun-if-changed=dom/webidls");
+    println!("cargo::rerun-if-changed=dom/bindings/codegen");
+    println!("cargo::rerun-if-changed={}", css_properties_json.display());
+    // FIXME: rerun-if-changed third_party/{ply, WebIDL}
```

Comment:
Is there a reason this change cannot be made now?

Reply:
because we would need to go in parent directory and I am not sure cargo would be happy about this (this would also be change from current behavior). I think it would be more appropriate to just move ply and webidl back to script as this would also make vendoring work more properly: https://github.com/servo/servo/issues/31812.

Reply:
1. Moving the vendored sources is a different issue than this one and it isn't so simple as `cargo vendor` also requires us to figure out what to do with the `resources` directory. Do we event want to support `cargo vendor`?
2. Identifying a file from a different directory works fine (or else the dependency on the CSS properties file wouldn't work). For instance, I've tested this locally and this seems to work fine

```diff
diff --git a/components/script/build.rs b/components/script/build.rs
index b740c53364..3ab015504f 100644
--- a/components/script/build.rs
+++ b/components/script/build.rs
@@ -21,8 +21,8 @@ fn main() {
 
     println!("cargo::rerun-if-changed=dom/webidls");
     println!("cargo::rerun-if-changed=dom/bindings/codegen");
+    println!("cargo::rerun-if-changed=../../third_party/WebIDL/WebIDL.py");
     println!("cargo::rerun-if-changed={}", css_properties_json.display());
-    // FIXME: rerun-if-changed third_party/{ply, WebIDL}
 
     let status = Command::new(find_python())
         .arg("dom/bindings/codegen/run.py")
```

This is an improvement to what we have now and is better than adding a `TODO` which may never be solved. Note that identifying the `ply` directory here didn't work well, likely because `cargo` is picking up changes to Python compiled code caches. I think it isn't as important to identify `ply` anyway. Let's integrate this into the PR if it looks good to you.

Reply:
> Identifying a file from a different directory works fine (or else the dependency on the CSS properties file wouldn't work)

Indeed.

> Note that identifying the ply directory here didn't work well, likely because cargo is picking up changes to Python compiled code caches

there is also pycache in webidl folder, maybe we need to include only python files, will investigate this.

> I think it isn't as important to identify ply anyway

Indeed, as we do not patch it.

Reply:
> Moving the vendored sources is a different issue than this one and it isn't so simple as cargo vendor also requires us to figure out what to do with the resources directory. Do we event want to support cargo vendor?

Do we want to support cargo vendor? Probably. For the resource directory we could wrap it in crate that would provide all files, but this is indeed separate issue.

Reply:
> there is also pycache in webidl folder, maybe we need to include only python files, will investigate this.

Ah, you already only included only webidl.py.